java/arcs/core/storage/StorageProxy.kt

```diff
@@ -93,19 +100,33 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperationAtTime, T>(
             .buffer()
             .onEach {
                 // TODO(jasonwyatt): Make the deferred lazy, so that we only idle when the client
-                //  requests it.
+                //  requests it, we could probably just use the busySendingMessagesChannel thinger.
+
                 val success = withTimeoutOrNull(5000) { store.idle() }
                 if (success == null) {
                     log.warning {
                         "Timeout exceeded (5 seconds) while waiting for store to become idle."
                     }
                 }
                 it.complete(true)
+
+                val queueLength = outgoingMessagesInFlight.getAndDecrement()
+                log.debug { "Store went idle with outgoing queue length: $queueLength" }
+                if (queueLength == 1) {
+                    busySendingMessagesChannel.send(false)
+                }
             }
             .flowOn(Dispatchers.Default)
             .launchIn(scheduler.scope)
     }
 
+    /**
+     * Suspends the coroutine while the [store] is busy processing our "outgoing messages".
+     */
+    suspend fun awaitOutgoingMessageQueueDrain() {
+        busySendingMessagesChannel.asFlow().debounce(50).filter { !it }.first()
```

Comment:
Since this is a ConflatedBroadcastChannel I think this will only debounce if the queue gets drained multiple times in the 50 ms window.
If that property is important for tests passing, it sounds like a possibility of introducing flakiness should the test runner act slowly.

Reply:
It's not important for the tests, but I felt it was important for an actual idleness-signal in general. (ie. it should be not-busy for at least 50ms before we say it went idle)

Reply:
Also, I ran the tests 100,000 times.. so I'm pretty sure it's not flaky :D